Feature/spi nor flash wp hold pins

JPHutchins opened the pull request:
The [jedec,spi-nor compatible](https://docs.zephyrproject.org/latest/build/dts/api/bindings/mtd/jedec,spi-nor.html) defines the use of wp-gpios (Write Protect) and hold-gpios (Hold) pins.  Only one of the flash drivers, `spi_flash_at45.c`, initializes and uses wp-gpios and none of them initialize or use hold-gpios.

This PR adds initialization and implementation of wp-gpios to the more general `spi_nor.c` driver and initializes but does not implement hold-gpios (set to inactive).

Two build tests were added and they can be collected with `west twister -c -p nrf52840dk_nrf52840 -T tests/drivers/flash`.

~~`spi_nor.c` has been formatted with clang-format, so it may be helpful to look at the functional changes only in https://github.com/zephyrproject-rtos/zephyr/commit/d551e5b71a36dbcb654f6dadd482682017f21bf8.~~

Discussions: 
* https://devzone.nordicsemi.com/f/nordic-q-a/99062/external-flash-write-protect-and-hold-pins
* https://devzone.nordicsemi.com/f/nordic-q-a/89971/write-protection-pin-for-external-spi-flash-configuration/380516

JPHutchins committed:
drivers: flash: implement wp-gpios and hold-gpios

Adds the wp-gpios & hold-gpios from jedec,spi-nor compatible to spi_nor.c.

Signed-off-by: J.P. Hutchins <jp@intercreate.io>

JPHutchins committed:
test: drivers: flash: build spi_nor.c wp & hold

Test spi_nor.c driver builds with or without wp-gpios and hold-gpios.

Signed-off-by: J.P. Hutchins <jp@intercreate.io>

JPHutchins commented:
Hi @de-nordic , can you retrigger the "Run tests with twister / twister-build (8) (pull_request_target)" pipeline when you have chance, none of these changes should impact the failing test case.  Thanks!

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -1254,12 +1279,37 @@ static int spi_nor_configure(const struct device *dev)
 static int spi_nor_init(const struct device *dev)
 {
+#if (ANY_INST_HAS_WP_GPIOS || ANY_INST_HAS_HOLD_GPIOS)
+	const struct spi_nor_config *cfg = dev->config;
```
Maybe we should have some macro like SPI_NOR_CONFIG or under some other name, that would just return us the cast pointer to `const struct spi_nod_config` made of `dev->config`. I do not like so many ifdefs just to define variable.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -1254,12 +1279,37 @@ static int spi_nor_configure(const struct device *dev)
 		k_sem_init(&driver_data->sem, 1, K_SEM_MAX_LIMIT);
 	}
 
+#if ANY_INST_HAS_WP_GPIOS
+	if (cfg->wp) {
+		if (!device_is_ready(cfg->wp->port)) {
+			return -ENODEV;
+		}
+		if (gpio_pin_configure_dt(cfg->wp, GPIO_OUTPUT_ACTIVE)) {
+			return -ENODEV;
+		}
+	}
+#endif /* ANY_INST_HAS_WP_GPIOS */
+#if ANY_INST_HAS_HOLD_GPIOS
+	if (cfg->hold) {
+		if (!device_is_ready(cfg->hold->port)) {
+			return -ENODEV;
+		}
+		if (gpio_pin_configure_dt(cfg->hold, GPIO_OUTPUT_INACTIVE)) {
+			return -ENODEV;
+		}
+	}
+#endif /* ANY_INST_HAS_HOLD_GPIOS */
```
At least some LOG_DBG is needed here. There are now four (probably additional) ways for the code to return `-ENODEV`, which require engaging debugger to figure out.

JPHutchins replied:
@de-nordic It is an unexpected error state, would LOG_ERR be OK?

de-nordic replied:
OK, so there is no way for caller to report this, as `z_sys_init_run_level` seems to log nothing, this means that here is the only place the problem can be described by log; assuming that a device is full tested and some design problems have been weed out, this problems will only happen in field. This means LOG_DBG may require pulling board of the field or uploading DBG software, where uploading debug software may not be possible in case when DFU happens to external flash.
Yes. I think this should be LOG_ERR. Question is whether we can save some flash by reducing the LOG_ERR. Text does not compile that well.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -821,8 +837,17 @@ static int spi_nor_erase(const struct device *dev, off_t addr, size_t size)
+#endif
 	int ret;
 
+#if ANY_INST_HAS_WP_GPIOS
+	if (cfg->wp) {
+		gpio_pin_set_dt(cfg->wp, write_protect);
+	}
+#endif
+
```
We will need to figure this out, probably later, but right now it seems that we are placing two write locks for, probably, no reason.

JPHutchins replied:
@de-nordic If I understand correctly, the driver presently supports soft write-protect and this PR adds GPIO write-protect.  

Users that tie WP disabled in HW won't define wp-gpios, so this PR should not change anything for them.  Users that connect WP to their MCU and then define wp-gpios in DTS incorrectly assume it's doing something (as far as I can tell it doesn't even initialize) and this PR addresses that.

Perhaps we should consider that defining wp-gpios in the DTS should disable the existing soft write-protect in this driver in favor of the gpio write-protect?  Personally I am OK with redundant write-protect!

de-nordic replied:
Twice the reasons to have errors and extra code, but let it be as is.

JPHutchins replied:
I too dislike code!

de-nordic commented:
Please rebase the PR, there seems to be some problems, with CI, unrelated to the changes.

JPHutchins committed:
drivers: flash: add LOG_ERR for pin init

-ENODEV returned in 4 cases, logs can distinguish

Signed-off-by: J.P. Hutchins <jp@intercreate.io>

JPHutchins committed:
drivers: flash: cleanup #ifs; fix missing ;

Fixes made running tests/drivers/flash

Signed-off-by: J.P. Hutchins <jp@intercreate.io>

JPHutchins commented:
@de-nordic @nashif LMK if you have any questions!

JPHutchins commented on drivers/flash/spi_nor.c:
```diff
@@ -823,6 +841,12 @@ static int spi_nor_write_protection_set(const struct device *dev,
 {
 	int ret;
 
+#if ANY_INST_HAS_WP_GPIOS
+	if (DEV_CFG(dev)->wp) {
+		gpio_pin_set_dt(DEV_CFG(dev)->wp, write_protect);
+	}
+#endif
+
```
Yes!  I do run this patch on a board with external SPI NOR.  My patch is either slightly different or I am lucky that it's working at all...  I suppose that manufacturers could allow config writes while write protect is enabled, but the proposed change seems much safer.

JPHutchins committed:
drivers: flash: fix hw write protect before sw

This change sets write-protect pin disabled BEFORE SW write-protect disable
and  write-protect pin enabled AFTER SW write-protect enable.

Signed-off-by: J.P. Hutchins <jp@intercreate.io>

JPHutchins commented:
I am not certain how to address the pipeline failure.  I am going to rebase again and see if it was already fixed.

```
-- Found BOARD.dts: /__w/zephyr/zephyr/boards/arm/mimxrt1160_evk/mimxrt1160_evk_cm7.dts
devicetree error: 'bus-speed-data' is marked as required in 'properties:' in /__w/zephyr/zephyr/dts/bindings/can/nxp,flexcan-fd.yaml, but does not appear in <Node /soc/can@40c3c000 in '/__w/zephyr/zephyr/misc/empty_file.c'>
-- In: /__w/zephyr/zephyr/twister-out/mimxrt1160_evk_cm7/tests/drivers/flash/common/drivers.flash.common.default/zephyr, command: /usr/bin/python3.10;/__w/zephyr/zephyr/scripts/dts/gen_defines.py;--dts;/__w/zephyr/zephyr/twister-out/mimxrt1160_evk_cm7/tests/drivers/flash/common/drivers.flash.common.default/zephyr/zephyr.dts.pre;--dtc-flags;'';--bindings-dirs;/__w/zephyr/zephyr/dts/bindings;--header-out;/__w/zephyr/zephyr/twister-out/mimxrt1160_evk_cm7/tests/drivers/flash/common/drivers.flash.common.default/zephyr/include/generated/devicetree_generated.h.new;--dts-out;/__w/zephyr/zephyr/twister-out/mimxrt1160_evk_cm7/tests/drivers/flash/common/drivers.flash.common.default/zephyr/zephyr.dts.new;--edt-pickle-out;/__w/zephyr/zephyr/twister-out/mimxrt1160_evk_cm7/tests/drivers/flash/common/drivers.flash.common.default/zephyr/edt.pickle;--edtlib-Werror;--vendor-prefixes;/__w/zephyr/zephyr/dts/bindings/vendor-prefixes.txt
CMake Error at cmake/modules/dts.cmake:276 (message):
  gen_defines.py failed with return code: 1
Call Stack (most recent call first):
  cmake/modules/zephyr_default.cmake:115 (include)
  share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  cmake/package_helper.cmake:115 (find_package)
```
